File: bridge/file_mapper.cpp
/**
 * 8.3 DOS Filename Mapper & Sanitizer Implementation
 * Solves Long-Filename collisions for Turbo C++ 3.0
 * Credits: ENCRYPTED CREW
 */

#include "file_mapper.hpp"
#include <algorithm>
#include <cctype>
#include <set>

DosFileMapper& DosFileMapper::instance() {
    static DosFileMapper s_instance;
    return s_instance;
}

DosFileMapper::DosFileMapper() {
    // Standard Borland files
    m_host_to_dos["HELLO.CPP"] = "HELLO.CPP";
    m_dos_to_host["HELLO.CPP"] = "HELLO.CPP";
}

bool DosFileMapper::is_dos_reserved(const std::string& name_no_ext) {
    static const std::set<std::string> reserved = {
        "CON", "PRN", "AUX", "NUL",
        "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
        "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
    };
    std::string upper = name_no_ext;
    std::transform(upper.begin(), upper.end(), upper.begin(), ::toupper);
    return reserved.find(upper) != reserved.end();
}

bool DosFileMapper::is_safe_relative_path(const std::string& relative_path) {
    if (relative_path.empty()) return false;
    if (relative_path.find("..") != std::string::npos) return false;
    if (relative_path[0] == '/' || relative_path[0] == '\\') return false;
    return true;
}

std::string DosFileMapper::to_dos_name(const std::string& host_name, const std::vector<std::string>& existing_names) {
    std::lock_guard<std::mutex> lock(m_mutex);

    auto it = m_host_to_dos.find(host_name);
    if (it != m_host_to_dos.end()) {
        return it->second;
    }

    size_t dot_pos = host_name.find_last_of('.');
    std::string base = (dot_pos != std::string::npos) ? host_name.substr(0, dot_pos) : host_name;
    std::string ext = (dot_pos != std::string::npos) ? host_name.substr(dot_pos + 1) : "";

    // Sanitize to uppercase alphanumeric
    std::string clean_base;
    for (char c : base) {
        if (std::isalnum(c) || c == '_' || c == '$' || c == '-') {
            clean_base += static_cast<char>(std::toupper(c));
        } else {
            clean_base += '_';
        }
    }
    if (clean_base.empty()) clean_base = "FILE";

    std::string clean_ext;
    for (char c : ext) {
        if (std::isalnum(c)) clean_ext += static_cast<char>(std::toupper(c));
    }
    if (clean_ext.size() > 3) clean_ext = clean_ext.substr(0, 3);

    if (is_dos_reserved(clean_base)) {
        clean_base = "X_" + clean_base.substr(0, 6);
    }

    std::set<std::string> existing_set;
    for (const auto& name : existing_names) {
        std::string upper = name;
        std::transform(upper.begin(), upper.end(), upper.begin(), ::toupper);
        existing_set.insert(upper);
    }

    std::string candidate_base = clean_base.substr(0, 8);
    std::string candidate_dos = clean_ext.empty() ? candidate_base : (candidate_base + "." + clean_ext);
    int collision_index = 1;

    while (existing_set.find(candidate_dos) != existing_set.end()) {
        std::string suffix = "~" + std::to_string(collision_index);
        size_t allowed_len = (suffix.size() < 8) ? (8 - suffix.size()) : 1;
        candidate_base = clean_base.substr(0, allowed_len) + suffix;
        candidate_dos = clean_ext.empty() ? candidate_base : (candidate_base + "." + clean_ext);
        collision_index++;
    }

    m_host_to_dos[host_name] = candidate_dos;
    m_dos_to_host[candidate_dos] = host_name;

    return candidate_dos;
}

std::string DosFileMapper::to_host_name(const std::string& dos_name) {
    std::lock_guard<std::mutex> lock(m_mutex);
    std::string upper = dos_name;
    std::transform(upper.begin(), upper.end(), upper.begin(), ::toupper);

    auto it = m_dos_to_host.find(upper);
    if (it != m_dos_to_host.end()) {
        return it->second;
    }
    return dos_name;
}

File: src/utils/fileMapper.ts
/**
 * Filename Mapper & Sanitizer for DOS 8.3 Filesystems
 * Ported from /bridge/file_mapper.cpp for the client-side host shell
 * Credits: ENCRYPTED CREW
 */

import { DosFile, FilenameMapping } from '../types/emulator';

const DOS_RESERVED_NAMES = new Set([
  'CON', 'PRN', 'AUX', 'NUL',
  'COM1', 'COM2', 'COM3', 'COM4', 'COM5', 'COM6', 'COM7', 'COM8', 'COM9',
  'LPT1', 'LPT2', 'LPT3', 'LPT4', 'LPT5', 'LPT6', 'LPT7', 'LPT8', 'LPT9'
]);

export class DosFileMapper {
  private mappings: Map<string, FilenameMapping> = new Map();
  private reverseMappings: Map<string, string> = new Map();

  constructor() {
    this.loadInitialMappings();
  }

  private loadInitialMappings() {
    // Pre-mapped standard files
    this.registerMapping('HELLO.CPP', 'HELLO.CPP');
    this.registerMapping('RAINBOW.C', 'RAINBOW.C');
    this.registerMapping('COLOURS.CPP', 'COLOURS.CPP');
    this.registerMapping('SOUND.CPP', 'SOUND.CPP');
    this.registerMapping('CIRCLE.CPP', 'CIRCLE.CPP');
  }

  /**
   * Sanitizes a host filename into a valid DOS 8.3 name.
   * Format: [BASENAME 1-8 chars].[EXT 0-3 chars], all uppercase, no illegal symbols.
   */
  public toDosName(hostName: string, existingDosNames: string[] = []): string {
    const cleanHostName = hostName.trim();
    if (this.mappings.has(cleanHostName)) {
      return this.mappings.get(cleanHostName)!.dosName;
    }

    // Path traversal check
    if (cleanHostName.includes('..') || cleanHostName.includes('/') || cleanHostName.includes('\\')) {
      throw new Error(`Security Violation: Path traversal detected in filename '${cleanHostName}'`);
    }

    const lastDotIndex = cleanHostName.lastIndexOf('.');
    let base = lastDotIndex !== -1 ? cleanHostName.substring(0, lastDotIndex) : cleanHostName;
    let ext = lastDotIndex !== -1 ? cleanHostName.substring(lastDotIndex + 1) : '';

    // Sanitize base: allow only A-Z, 0-9, and valid DOS chars
    base = base.toUpperCase().replace(/[^A-Z0-9_$-]/g, '_');
    ext = ext.toUpperCase().replace(/[^A-Z0-9]/g, '').substring(0, 3);

    if (base.length === 0) {
      base = 'FILE';
    }

    // Check if reserved DOS name
    if (DOS_RESERVED_NAMES.has(base)) {
      base = `X_${base.substring(0, 6)}`;
    }

    let candidateBase = base.substring(0, 8);
    let candidateDosName = ext.length > 0 ? `${candidateBase}.${ext}` : candidateBase;

    // Handle collision if already used by another file
    const existingSet = new Set(existingDosNames.map(n => n.toUpperCase()));
    let collisionNum = 1;

    while (existingSet.has(candidateDosName)) {
      const suffix = `~${collisionNum}`;
      const allowedBaseLen = Math.max(1, 8 - suffix.length);
      candidateBase = `${base.substring(0, allowedBaseLen)}${suffix}`;
      candidateDosName = ext.length > 0 ? `${candidateBase}.${ext}` : candidateBase;
      collisionNum++;

      if (collisionNum > 999) {
        throw new Error(`Exceeded maximum collisions for 8.3 filename: ${hostName}`);
      }
    }

    this.registerMapping(cleanHostName, candidateDosName, collisionNum - 1);
    return candidateDosName;
  }

  public getHostName(dosName: string): string {
    return this.reverseMappings.get(dosName.toUpperCase()) || dosName;
  }

  public registerMapping(hostName: string, dosName: string, collisionIndex: number = 0) {
    const mapping: FilenameMapping = {
      hostName,
      dosName: dosName.toUpperCase(),
      collisionIndex,
      createdAt: Date.now()
    };
    this.mappings.set(hostName, mapping);
    this.reverseMappings.set(dosName.toUpperCase(), hostName);
  }

  public getAllMappings(): FilenameMapping[] {
    return Array.from(this.mappings.values());
  }

  /**
   * Preserves CRLF line endings required by Borland Turbo C++
   */
  public static ensureDosCrlf(text: string): string {
    // Normalize to LF first, then replace LF with CRLF
    const normalized = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
    return normalized.replace(/\n/g, '\r\n');
  }
}

export const globalFileMapper = new DosFileMapper();
